python/ray/actor.py

```diff
@@ -40,31 +42,130 @@ def get_actor_method_function_id(attr):
+
+
+def put_dummy_object(worker, dummy_object_id):
+    """Put a dummy actor object into the local object store.
+
+    This registers a dummy object ID in the local store with an empty numpy
+    array as the value. The resulting object is pinned to the store by storing
+    it to the worker's state.
+
+    For actors, dummy objects are used to store the stateful dependencies
+    between consecutive method calls. This function should be called for every
+    actor method execution that updates the actor's internal state.
+
+    Args:
+        worker: The worker to use to perform the put.
+        dummy_object_id: The object ID of the dummy object.
+
+    """
+    # Add the dummy output for actor tasks. TODO(swang): We use
+    # a numpy array as a hack to pin the object in the object
+    # store.  Once we allow object pinning in the store, we may
+    # use `None`.
+    dummy_object = np.zeros(1)
+    worker.put_object(dummy_object_id, dummy_object)
+    # Keep the dummy output in scope for the lifetime of the
+    # actor, to prevent eviction from the object store.
+    dummy_object = worker.get_object([dummy_object_id])
+    dummy_object = dummy_object[0]
+    worker.actor_pinned_objects[dummy_object_id] = dummy_object
+
+
+def is_checkpoint_task(task_counter, checkpoint_interval):
+    if checkpoint_interval <= 0:
+        return False
+    return (task_counter % checkpoint_interval == 0)
+
+
+def make_actor_method_executor(worker, method_name, method):
+    """Make an executor that wraps a user-defined actor method.
+
+    The executor wraps the method to update the worker's internal state. If the
+    task is a success, the dummy object returned is added to the object store,
+    to signal that the following task can run, and the worker's task counter is
+    updated to match the executed task. Else, the executor reports failure to
+    the local scheduler so that the task counter does not get updated.
+
+    Args:
+        worker (Worker): The worker that is executing the actor.
+        method_name (str): The name of the actor method.
+        method (instancemethod): The actor method to wrap. This should be a
+            method defined on the actor class and should therefore take an
+            instance of the actor as the first argument.
+
+    Returns:
+        A function that executes the given actor method on the worker's stored
+        instance of the actor. The function also updates the worker's internal
+        state to record the executed method.
+    """
+
+    def actor_method_executor(dummy_return_id, task_counter, actor,
+                              *args):
+        # An actor task's dependency on the previous task is represented by
+        # a dummy argument. Remove this argument before invocation.
+        args = args[:-1]
+        if method_name == "__ray_checkpoint__":
+            # Execute the checkpoint task. NOTE(swang): Checkpoint methods
+            # should not throw an exception.
```

Comment:
They definitely could throw an exception, right? E.g., if they try to pickle stuff, pickling can fail. Also, if we allow user-defined checkpointing methods (which may be necessary for neural nets and things like that), the user could have a bug in the checkpointing method.

Does that have any implications for this code? Maybe we should add a test case for the case where checkpointing fails?

Reply:
Hmm what do you think the expected behavior should be if the checkpoint task does throw an exception? For both saving and resuming a checkpoint?

For me, I would probably expect saving a checkpoint to store the exception like a normal task, and then continue executing tasks on the actor. An exception while resuming a checkpoint could reconstruct the previous object, as if there were no successful checkpoints. How does that sound?

Yeah, once we figure out the expected behavior, a test case would be good. Right now a checkpoint task that throws an exception would just hang, since it would never put the dummy object.